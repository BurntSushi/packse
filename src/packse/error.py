import textwrap
from pathlib import Path


class PackseError(Exception):
    pass


class UserError(PackseError):
    pass


class DestinationAlreadyExists(UserError):
    def __init__(self, destination: Path) -> None:
        message = f"Destination directory '{destination}' already exists"
        super().__init__(message)


class InvalidScenario(UserError):
    def __init__(self, path: Path) -> None:
        message = f"File at '{path}' is not a valid scenario"
        super().__init__(message)


class ScenarioNotFound(UserError):
    def __init__(self, path: Path) -> None:
        message = f"Scenario '{path}' not found"
        super().__init__(message)


class BuildError(PackseError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)
