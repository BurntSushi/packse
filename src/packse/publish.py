"""
Publish package distributions.
"""
import logging
import subprocess
import textwrap
from pathlib import Path

from packse.error import (
    InvalidPublishTarget,
    PublishAlreadyExists,
    PublishError,
)

logger = logging.getLogger(__name__)


def publish(targets: list[Path], skip_existing: bool = True):
    for target in targets:
        if not target.is_dir():
            raise InvalidPublishTarget(target, reason="Not a directory.")

    logger.info("Publishing %s targets...", len(targets))
    for target in targets:
        logger.info("Publishing '%s'...", target.name)
        for distfile in target.iterdir():
            try:
                publish_package_distribution(distfile)
            except PublishAlreadyExists:
                if not skip_existing:
                    raise
                logger.info("Skipped '%s': already published.", distfile.name)
            else:
                logger.info("Published '%s'", distfile.name)


def publish_package_distribution(target: Path) -> None:
    """
    Publish package distribution.
    """
    try:
        output = subprocess.check_output(
            ["twine", "upload", "-r", "testpypi", str(target.resolve())],
            stderr=subprocess.STDOUT,
        )
    except subprocess.CalledProcessError as exc:
        output = exc.output.decode()
        if "File already exists" in output:
            raise PublishAlreadyExists(target.name)
        raise PublishError(
            f"Publishing {target} with twine failed",
            output,
        )
    else:
        logger.debug(
            "Publishing %s:\n\n%s",
            target.name,
            textwrap.indent(output.decode(), " " * 4),
        )
