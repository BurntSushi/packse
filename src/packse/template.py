import json
import logging
from pathlib import Path
from typing import Any

from packse.templates import __templates_path__

TEMPLATE_VERSION = ".template-version"

logger = logging.getLogger(__name__)


def get_template_version(template_name: str) -> str:
    template_path = __templates_path__ / template_name
    version_path = template_path / TEMPLATE_VERSION
    if version_path.exists():
        return version_path.read_text()
    else:
        return "0"


def create_from_template(
    destination: Path, template_name: str, variables: dict[str, Any]
) -> Path:
    template_path = __templates_path__ / template_name
    first_root = None
    for root, _, files in template_path.walk():
        # Determine the new directory path in the destination
        new_root = destination / Path(
            replace_placeholders(str(root), variables)
        ).relative_to(template_path)

        if new_root == destination:
            continue

        if not first_root:
            first_root = new_root

        # Create the new directory
        logger.info("Creating %s", new_root.relative_to(destination))
        new_root.mkdir()

        for file in files:
            file_path = root / file

            # Determine the new file path
            new_file_path = new_root / replace_placeholders(file, variables)
            logger.info("Creating %s", new_file_path.relative_to(destination))

            new_file_path.write_text(
                replace_placeholders(file_path.read_text(), variables)
            )

    return first_root


def replace_placeholders(target: str, variables: dict[str, Any]):
    for key, value in variables.items():
        target = target.replace(f"{{{{ {key} }}}}", stringify(value))
    return target


def stringify(value: Any) -> str:
    if isinstance(value, str):
        return value
    else:
        # TOOD(zanieb): Consider inferring the format from the file extension
        return json.dumps(value)
