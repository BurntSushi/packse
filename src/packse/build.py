"""
Build packages for the given scenarios.
"""
import logging
import shutil
import subprocess
import textwrap
from pathlib import Path
from typing import Generator

from packse.error import (
    BuildError,
    DestinationAlreadyExists,
    InvalidScenario,
    ScenarioNotFound,
)
from packse.scenario import Package, Scenario, load_scenario, scenario_prefix
from packse.template import create_from_template

logger = logging.getLogger(__name__)


def build(targets: list[Path], rm_destination: bool):
    # Validate all targets first
    for target in targets:
        if not target.exists():
            raise ScenarioNotFound(target)

        try:
            load_scenario(target)
        except Exception as exc:
            raise InvalidScenario(target, reason=str(exc)) from exc

    # Then build each one
    for target in targets:
        result = build_scenario(target, rm_destination)
        print(result)


def build_scenario(target: Path, rm_destination: bool) -> str:
    """
    Build the scenario defined at the given path.

    Returns the scenario's root package name.
    """
    scenario = load_scenario(target)
    prefix = scenario_prefix(scenario)

    work_dir = Path.cwd()
    build_destination = work_dir / "build" / prefix
    dist_destination = work_dir / "dist" / prefix

    logging.info(
        "Building '%s' in directory '%s'",
        prefix,
        build_destination.relative_to(work_dir),
    )

    if build_destination.exists():
        if rm_destination:
            shutil.rmtree(build_destination)
        else:
            raise DestinationAlreadyExists(build_destination)

    build_destination.mkdir(parents=True)

    if dist_destination.exists():
        if rm_destination:
            shutil.rmtree(dist_destination)
        else:
            raise DestinationAlreadyExists(dist_destination)

    dist_destination.mkdir(parents=True)

    for name, package in scenario.packages.items():
        build_scenario_package(
            scenario=scenario,
            prefix=prefix,
            name=name,
            package=package,
            work_dir=work_dir,
            build_destination=build_destination,
            dist_destination=dist_destination,
        )

    return f"{prefix}-{scenario.root}"


def build_scenario_package(
    scenario: Scenario,
    prefix: str,
    name: str,
    package: Package,
    work_dir: Path,
    build_destination: Path,
    dist_destination: Path,
):
    package_name = f"{prefix}-{name}"

    # Generate a Python module name
    module_name = package_name.replace("-", "_")

    for version, specification in package.versions.items():
        package_destination = create_from_template(
            build_destination,
            template_name=scenario.template,
            variables={
                "scenario-name": scenario.name,
                "package-name": package_name,
                "module-name": module_name,
                "version": version,
                "dependencies": [f"{prefix}-{spec}" for spec in specification.requires],
                "requires-python": specification.requires_python,
            },
        )

        logger.info(
            "Building %s with hatch",
            package_destination.relative_to(work_dir),
        )

        for dist in build_package_distributions(package_destination):
            shared_path = dist_destination / dist.name
            logger.info("Linked distribution to %s", shared_path.relative_to(work_dir))
            shared_path.hardlink_to(dist)


def build_package_distributions(target: Path) -> Generator[Path, None, None]:
    """
    Build package distributions, yield each built distribution path, then delete the distribution folder.
    """
    try:
        output = subprocess.check_output(
            ["hatch", "build"],
            cwd=target,
            stderr=subprocess.STDOUT,
        )

        yield from sorted((target / "dist").iterdir())
        shutil.rmtree(target / "dist")

    except subprocess.CalledProcessError as exc:
        raise BuildError(
            f"Building {target.name} with hatch failed",
            exc.output.decode(),
        )
    else:
        logger.debug(
            "Building %s:\n\n%s",
            target.name,
            textwrap.indent(output.decode(), " " * 4),
        )
