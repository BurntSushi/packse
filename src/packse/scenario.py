import hashlib
import os
from pathlib import Path

import msgspec

from packse.template import get_template_version


class PackageVersion(msgspec.Struct):
    requires_python: str | None = ">=3.7"
    requires: list[str] = []
    sdist: bool = True
    wheel: bool = True
    description: str = ""

    def hash(self) -> str:
        """
        Return a hash of the contents
        """
        hasher = hashlib.new("md5", usedforsecurity=False)
        hasher.update((self.requires_python or "").encode())
        for require in self.requires:
            hasher.update(require.encode())
        return hasher.hexdigest()


class Package(msgspec.Struct):
    versions: dict[str, PackageVersion]

    def hash(self) -> str:
        """
        Return a hash of the contents
        """
        hasher = hashlib.new("md5", usedforsecurity=False)
        for name, version in self.versions.items():
            hasher.update(name.encode())
            hasher.update(version.hash().encode())
        return hasher.hexdigest()


class Scenario(msgspec.Struct):
    name: str
    """
    The name of the scenario.
    """

    packages: dict[str, Package]
    """
    The packages available in the scenario.
    """

    root: str
    """
    The root package of the scenario.

    The scenario entrypoint package will require this package.
    """

    template: str = "simple"
    """
    The template to use for scenario packages.
    """

    description: str | None = None
    """
    The description of the scenario.
    """

    def hash(self) -> str:
        """
        Return a hash of the scenario contents
        """
        hasher = hashlib.new("md5", usedforsecurity=False)
        hasher.update(self.name.encode())
        hasher.update(self.template.encode())
        for name, package in self.packages.items():
            hasher.update(name.encode())
            hasher.update(package.hash().encode())
        return hasher.hexdigest()


def load_scenario(target: Path) -> Scenario:
    """
    Loads a scenario
    """
    return msgspec.json.decode(target.read_text(), type=Scenario)


def load_many_scenarios(target: Path) -> list[Scenario]:
    """
    Loads a file with many scenarios
    """
    return msgspec.json.decode(target.read_text(), type=list[Scenario])


def load_scenarios(target: Path) -> list[Scenario]:
    # Guess if the file contains one or many scenario
    with target.open() as buffer:
        many = buffer.readline().lstrip().startswith("[")
    if many:
        return load_many_scenarios(target)
    else:
        return [load_scenario(target)]


def scenario_version(scenario: Scenario) -> str:
    """
    Generate a unique version for a scenario based on its contents.
    """
    template_version = get_template_version(scenario.template)
    hasher = hashlib.new("md5", usedforsecurity=False)
    hasher.update(template_version.encode())
    hasher.update(scenario.hash().encode())
    hasher.update(os.environ.get("PACKSE_VERSION_SEED", "").encode())
    return hasher.hexdigest()[:8]


def scenario_prefix(scenario: Scenario) -> str:
    """
    Generate a prefix for a scenario.
    """
    version = scenario_version(scenario)
    return f"{scenario.name}-{version}"
