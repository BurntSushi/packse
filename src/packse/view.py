"""
View package tree for the given scenarios.
"""
import logging
from pathlib import Path

from packaging.requirements import Requirement

from packse.error import InvalidScenario, ScenarioNotFound
from packse.scenario import Scenario, load_scenarios, scenario_prefix

logger = logging.getLogger(__name__)


def view(targets: list[Path]):
    scenarios = []

    # Validate and collect all targets first
    for target in targets:
        if not target.exists():
            raise ScenarioNotFound(target)

        try:
            logger.debug("Loading %s", target)
            scenarios.extend(load_scenarios(target))
        except Exception as exc:
            raise InvalidScenario(target, reason=str(exc)) from exc

    # Then view each one
    for scenario in scenarios:
        logger.debug("Viewing %s", scenario.name)
        view_scenario(scenario)


def view_scenario(scenario: Scenario):
    prefix = scenario_prefix(scenario)

    print(prefix)
    print(dependency_tree(scenario))


def dependency_tree(scenario: Scenario):
    """
    Generate a dependency tree for a scenario
    """

    space = "    "
    branch = "│   "
    tee = "├── "
    last = "└── "
    buffer = ""

    def render_package_versions(
        package: str,
        prefix: str = "",
        for_requirement: Requirement | None = None,
    ):
        versions = scenario.packages[package].versions

        pointers = [tee] * (len(versions) - 1) + [last]
        satisfied = False
        for pointer, version in zip(pointers, versions):
            if for_requirement and not for_requirement.specifier.contains(version):
                continue

            satisfied = True
            message = "satisfied by " if for_requirement else ""
            yield prefix + pointer + message + f"{package}-{version}"

            extension = branch if pointer == tee else space
            yield from render_requirements_for(
                package, version, prefix=prefix + extension
            )

        if for_requirement and not satisfied:
            yield prefix + last + "unsatisfied"

    def render_requirements_for(package: str, version: str, prefix: str = ""):
        requirements = list(scenario.packages[package].versions[version].requires)

        pointers = [tee] * (len(requirements) - 1) + [last]
        for pointer, requirement in zip(pointers, sorted(requirements)):
            yield prefix + pointer + "requires " + requirement

            parsed_requirement = Requirement(requirement)
            if parsed_requirement.name in scenario.packages:
                extension = branch if pointer == tee else space
                yield from render_package_versions(
                    parsed_requirement.name,
                    prefix=prefix + extension,
                    for_requirement=parsed_requirement,
                )
            else:
                yield prefix + space + last + "unsatisfied"

    for line in render_package_versions(scenario.root):
        buffer += line + "\n"

    for package in sorted(scenario.packages):
        if package == scenario.root:
            continue
        for line in render_package_versions(package):
            buffer += line + "\n"

    return buffer
