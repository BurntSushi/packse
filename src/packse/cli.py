import argparse
import logging
import sys
from pathlib import Path

from packse.build import build
from packse.error import BuildError, DestinationAlreadyExists, UserError
from packse.view import view


def entrypoint():
    parser = get_parser()
    args = parser.parse_args()
    if not hasattr(args, "call"):
        parser.print_help()
        return None

    if args.quiet:
        log_level = logging.CRITICAL
    elif args.verbose:
        log_level = logging.DEBUG
    else:
        log_level = logging.INFO

    logging.basicConfig(level=log_level)

    # Call the implementation wrapper e.g. `_call_build`
    try:
        args.call(args)
    except DestinationAlreadyExists as exc:
        print(f"{exc}. Pass `--rm` to allow removal.", file=sys.stderr)
        exit(1)
    except UserError as exc:
        print(f"{exc}.", file=sys.stderr)
        exit(1)
    except BuildError as exc:
        print(f"{exc}", file=sys.stderr)
        exit(1)


def _call_build(args):
    build(args.targets, rm_destination=args.rm)


def _call_view(args):
    view(args.targets)


def _root_parser():
    parser = argparse.ArgumentParser(
        description="Utilities for working example packaging scenarios",
    )
    _add_shared_arguments(parser)
    return parser


def _add_build_parser(subparsers):
    parser = subparsers.add_parser("build", help="Build packages for a scenario")
    parser.set_defaults(call=_call_build)
    parser.add_argument(
        "targets",
        type=Path,
        nargs="+",
        help="The scenario to build",
    )
    parser.add_argument(
        "--rm",
        action="store_true",
        help="Allow removal of existing build directory",
    )
    _add_shared_arguments(parser)


def _add_view_parser(subparsers):
    parser = subparsers.add_parser("view", help="View a scenario")
    parser.set_defaults(call=_call_view)
    parser.add_argument(
        "targets",
        type=Path,
        nargs="+",
        help="The scenario to view",
    )

    _add_shared_arguments(parser)


def _add_shared_arguments(parser):
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Enable debug logging",
    )

    parser.add_argument(
        "-q",
        "--quiet",
        action="store_true",
        help="Disable logging",
    )


def get_parser() -> argparse.ArgumentParser:
    parser = _root_parser()
    subparsers = parser.add_subparsers(title="commands")
    _add_build_parser(subparsers)
    _add_view_parser(subparsers)

    return parser
