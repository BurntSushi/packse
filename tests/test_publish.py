import pytest
import subprocess

from packse import __development_base_path__
from .common import snapshot_command
from pathlib import Path


def test_publish_example(snapshot, tmpcwd: Path):
    target = __development_base_path__ / "scenarios" / "example.json"

    # Build first
    # TODO(zanieb): Since we're doing a dry run consider just constructing some fake files?
    subprocess.check_call(
        ["packse", "build", str(target)],
        cwd=tmpcwd,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    dists = list((tmpcwd / "dist").iterdir())
    assert len(dists) == 1
    dist = dists[0]

    assert snapshot_command(["publish", "--dry-run", dist]) == snapshot
