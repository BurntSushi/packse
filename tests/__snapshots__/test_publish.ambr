# serializer version: 1
# name: test_publish_example
  dict({
    'exit_code': 0,
    'stderr': '''
      INFO:packse.publish:Publishing 1 targets...
      INFO:packse.publish:Publishing 'example-cd797223'...
      INFO:packse.publish:Published 'example_cd797223_b-1.0.0.tar.gz'
      INFO:packse.publish:Published 'example_cd797223_a-1.0.0.tar.gz'
      INFO:packse.publish:Published 'example_cd797223_a-1.0.0-py3-none-any.whl'
      INFO:packse.publish:Published 'example_cd797223_b-1.0.0-py3-none-any.whl'
  
    ''',
    'stdout': '''
      Would execute: twine upload -r testpypi $PWD/dist/example-cd797223/example_cd797223_b-1.0.0.tar.gz
      Would execute: twine upload -r testpypi $PWD/dist/example-cd797223/example_cd797223_a-1.0.0.tar.gz
      Would execute: twine upload -r testpypi $PWD/dist/example-cd797223/example_cd797223_a-1.0.0-py3-none-any.whl
      Would execute: twine upload -r testpypi $PWD/dist/example-cd797223/example_cd797223_b-1.0.0-py3-none-any.whl
  
    ''',
  })
# ---
